File: transports/rust/tests/conformance.rs
//! Verify the Rust reference library against the checked-in conformance vectors.
//!
//! This both guards the Rust encoders/decoders against drift and demonstrates the
//! contract every other language's reference library follows: reconstruct each
//! vector's envelope from its language-neutral `input`, assert encode → `hex`, and
//! assert decode(`hex`) → the same envelope.

use csilgen_transport::datagrams::*;
use csilgen_transport::events::*;
use csilgen_transport::rpc::*;
use serde_json::Value;

fn unhex(s: &str) -> Vec<u8> {
    (0..s.len())
        .step_by(2)
        .map(|i| u8::from_str_radix(&s[i..i + 2], 16).unwrap())
        .collect()
}

fn hex(bytes: &[u8]) -> String {
    bytes.iter().map(|b| format!("{b:02x}")).collect()
}

fn load(name: &str) -> Vec<Value> {
    let path = format!("{}/../conformance/{name}", env!("CARGO_MANIFEST_DIR"));
    let text = std::fs::read_to_string(&path).unwrap_or_else(|e| panic!("read {path}: {e}"));
    let doc: Value = serde_json::from_str(&text).unwrap();
    doc["vectors"].as_array().unwrap().clone()
}

fn opt_u64(v: &Value) -> Option<u64> {
    v.as_u64()
}

fn opt_str(v: &Value) -> Option<String> {
    v.as_str().map(|s| s.to_string())
}

#[test]
fn rpc_vectors() {
    for vec in load("rpc.json") {
        let name = vec["name"].as_str().unwrap();
        let input = &vec["input"];
        let expected_hex = vec["hex"].as_str().unwrap();
        let kind = input["kind"].as_str().unwrap();
        let actual = match kind {
            "request" => {
                let mut r = RpcRequest::new(
                    input["service"].as_str().unwrap(),
                    input["op"].as_str().unwrap(),
                    unhex(input["payload_hex"].as_str().unwrap()),
                );
                r.id = opt_u64(&input["id"]);
                r.auth = opt_str(&input["auth"]);
                let bytes = r.encode().unwrap();
                assert_eq!(RpcRequest::decode(&bytes).unwrap(), r, "decode {name}");
                bytes
            }
            "response" => {
                let r = RpcResponse {
                    id: opt_u64(&input["id"]),
                    status: csilgen_transport::Status::from_code(input["status"].as_i64().unwrap()),
                    variant: opt_str(&input["variant"]),
                    error: opt_str(&input["error"]),
                    payload: unhex(input["payload_hex"].as_str().unwrap()),
                };
                let bytes = r.encode().unwrap();
                assert_eq!(RpcResponse::decode(&bytes).unwrap(), r, "decode {name}");
                bytes
            }
            "push" => {
                let p = RpcPush::new(
                    input["service"].as_str().unwrap(),
                    input["event"].as_str().unwrap(),
                    unhex(input["payload_hex"].as_str().unwrap()),
                );
                let bytes = p.encode().unwrap();
                assert_eq!(RpcPush::decode(&bytes).unwrap(), p, "decode {name}");
                bytes
            }
            other => panic!("unknown rpc kind {other}"),
        };
        assert_eq!(hex(&actual), expected_hex, "encode {name}");
    }
}

#[test]
fn events_vectors() {
    for vec in load("events.json") {
        let name = vec["name"].as_str().unwrap();
        let input = &vec["input"];
        let expected_hex = vec["hex"].as_str().unwrap();

        // Control payloads are keyed by "control"; application events by "profile".
        if let Some(control) = input["control"].as_str() {
            let bytes = match control {
                "hello" => Hello {
                    versions: input["versions"]
                        .as_array()
                        .unwrap()
                        .iter()
                        .map(|x| x.as_u64().unwrap())
                        .collect(),
                    profiles: input["profiles"]
                        .as_array()
                        .unwrap()
                        .iter()
                        .map(|x| x.as_str().unwrap().to_string())
                        .collect(),
                    service: opt_str(&input["service"]),
                    auth: opt_str(&input["auth"]),
                }
                .encode()
                .unwrap(),
                "hello_ack" => HelloAck {
                    v: input["v"].as_u64().unwrap(),
                    profile: input["profile"].as_str().unwrap().to_string(),
                    session: opt_str(&input["session"]),
                }
                .encode()
                .unwrap(),
                "ping" => Heartbeat {
                    nonce: input["nonce"].as_u64().unwrap(),
                    at: opt_u64(&input["at"]),
                }
                .encode()
                .unwrap(),
                "close" => Close {
                    status: csilgen_transport::Status::from_code(input["status"].as_i64().unwrap()),
                    reason: opt_str(&input["reason"]),
                }
                .encode()
                .unwrap(),
                other => panic!("unknown control {other}"),
            };
            assert_eq!(hex(&bytes), expected_hex, "encode {name}");
            continue;
        }

        let profile = match input["profile"].as_str().unwrap() {
            "verbose" => Profile::Verbose,
            "compact" => Profile::Compact,
            other => panic!("unknown profile {other}"),
        };
        let payload = unhex(input["payload_hex"].as_str().unwrap());
        let ev = match profile {
            Profile::Verbose => {
                let mut e = Event::verbose(
                    opt_str(&input["service"]),
                    input["event"].as_str().unwrap(),
                    payload,
                );
                e.id = opt_u64(&input["id"]);
                e
            }
            Profile::Compact => {
                let mut e = Event::compact(
                    input["service_ord"].as_u64().unwrap(),
                    input["op_ord"].as_u64().unwrap(),
                    payload,
                );
                e.id = opt_u64(&input["id"]);
                e
            }
        };
        let bytes = ev.encode(profile).unwrap();
        assert_eq!(hex(&bytes), expected_hex, "encode {name}");
        assert_eq!(Event::decode(&bytes, profile).unwrap(), ev, "decode {name}");
    }
}

#[test]
fn datagram_vectors() {
    for vec in load("datagrams.json") {
        let name = vec["name"].as_str().unwrap();
        let input = &vec["input"];
        let expected_hex = vec["hex"].as_str().unwrap();
        match input["profile"].as_str().unwrap() {
            "cbor-array" => {
                let d = Datagram::new(
                    input["op_ord"].as_u64().unwrap(),
                    input["seq"].as_u64().unwrap(),
                    unhex(input["payload_hex"].as_str().unwrap()),
                );
                let bytes = d.encode().unwrap();
                assert_eq!(hex(&bytes), expected_hex, "encode {name}");
                assert_eq!(Datagram::decode(&bytes).unwrap(), d, "decode {name}");
            }
            "compact-header" => {
                let mut d = CompactDatagram::new(
                    input["op_ord"].as_u64().unwrap() as u8,
                    input["seq"].as_u64().unwrap() as u16,
                    unhex(input["body_hex"].as_str().unwrap()),
                );
                if let Some(e) = opt_u64(&input["epoch"]) {
                    d = d.with_epoch(e as u8);
                }
                let bytes = d.encode();
                assert_eq!(hex(&bytes), expected_hex, "encode {name}");
                assert_eq!(CompactDatagram::decode(&bytes).unwrap(), d, "decode {name}");
            }
            other => panic!("unknown datagram profile {other}"),
        }
    }
}

File: transports/ocaml/test/conformance.ml
(* Verify the OCaml reference library against the checked-in conformance vectors.
   Reconstruct each vector's envelope from its language-neutral [input], assert
   encode -> [hex], and assert decode([hex]) -> the same envelope. Mirrors
   transports/go/conformance_test.go and transports/rust/tests/conformance.rs.

   The vectors live at transports/conformance/, outside this dune project, so the
   directory is taken from CSILGEN_CONFORMANCE_DIR (set by run-tests.sh); a few
   relative fallbacks cover an in-tree run. *)

module J = Json
open Csilgen_transport

let read_file path =
  let ic = open_in_bin path in
  Fun.protect
    ~finally:(fun () -> close_in ic)
    (fun () -> really_input_string ic (in_channel_length ic))

let vectors_path name =
  let dir =
    match Sys.getenv_opt "CSILGEN_CONFORMANCE_DIR" with
    | Some d -> d
    | None -> (
        let candidates =
          [ "../conformance"; "../../conformance"; "../../../conformance" ]
        in
        match List.find_opt Sys.file_exists candidates with
        | Some d -> d
        | None -> "../conformance")
  in
  Filename.concat dir name

let load name =
  J.to_list (J.member "vectors" (J.parse (read_file (vectors_path name))))

let tohex b =
  let buf = Buffer.create (Bytes.length b * 2) in
  for i = 0 to Bytes.length b - 1 do
    Buffer.add_string buf (Printf.sprintf "%02x" (Bytes.get_uint8 b i))
  done;
  Buffer.contents buf

let unhex s =
  let n = String.length s / 2 in
  let b = Bytes.create n in
  for i = 0 to n - 1 do
    Bytes.set_uint8 b i (int_of_string ("0x" ^ String.sub s (i * 2) 2))
  done;
  b

let expect_hex name ~got ~want =
  if got <> want then (
    Printf.eprintf "FAIL %s encode:\n  got  %s\n  want %s\n" name got want;
    failwith ("conformance encode mismatch: " ^ name))

let expect name cond =
  if not cond then (
    Printf.eprintf "FAIL %s\n" name;
    failwith ("conformance mismatch: " ^ name))

(* Accessors over a vector's language-neutral [input] object. *)
let req_str input k = J.to_string (J.member k input)
let req_i64 input k = J.to_int64 (J.member k input)
let req_int input k = J.to_int (J.member k input)

let opt_str input k =
  match J.member k input with J.Str s -> Some s | _ -> None

let opt_i64 input k =
  match J.member k input with J.Num f -> Some (Int64.of_float f) | _ -> None

let opt_int input k =
  match J.member k input with J.Num f -> Some (int_of_float f) | _ -> None

let die name e = failwith (name ^ ": " ^ Conventions.error_message e)

let rpc_vectors () =
  List.iter
    (fun vec ->
      let name = J.to_string (J.member "name" vec) in
      let hex = J.to_string (J.member "hex" vec) in
      let input = J.member "input" vec in
      match req_str input "kind" with
      | "request" -> (
          let r =
            Rpc.new_request (req_str input "service") (req_str input "op")
              (unhex (req_str input "payload_hex"))
          in
          let r =
            match opt_i64 input "id" with
            | Some id -> Rpc.request_with_id r id
            | None -> r
          in
          let r =
            match opt_str input "auth" with
            | Some a -> Rpc.request_with_auth r a
            | None -> r
          in
          let bytes = Rpc.encode_request r in
          expect_hex name ~got:(tohex bytes) ~want:hex;
          match Rpc.decode_request bytes with
          | Ok dec -> expect (name ^ " decode") (dec = r)
          | Error e -> die (name ^ " decode") e)
      | "response" -> (
          let r : Rpc.response =
            {
              Rpc.id = opt_i64 input "id";
              status = Conventions.status_from_code (req_i64 input "status");
              variant = opt_str input "variant";
              error = opt_str input "error";
              payload = unhex (req_str input "payload_hex");
            }
          in
          let bytes = Rpc.encode_response r in
          expect_hex name ~got:(tohex bytes) ~want:hex;
          match Rpc.decode_response bytes with
          | Ok dec -> expect (name ^ " decode") (dec = r)
          | Error e -> die (name ^ " decode") e)
      | "push" -> (
          let p =
            Rpc.new_push (req_str input "service") (req_str input "event")
              (unhex (req_str input "payload_hex"))
          in
          let bytes = Rpc.encode_push p in
          expect_hex name ~got:(tohex bytes) ~want:hex;
          match Rpc.decode_push bytes with
          | Ok dec -> expect (name ^ " decode") (dec = p)
          | Error e -> die (name ^ " decode") e)
      | other -> failwith ("unknown rpc kind " ^ other))
    (load "rpc.json")

let events_vectors () =
  List.iter
    (fun vec ->
      let name = J.to_string (J.member "name" vec) in
      let hex = J.to_string (J.member "hex" vec) in
      let input = J.member "input" vec in
      let control = J.member "control" input in
      if not (J.is_null control) then
        let bytes =
          match J.to_string control with
          | "hello" ->
              let versions =
                List.map J.to_int64 (J.to_list (J.member "versions" input))
              in
              let profiles =
                List.map J.to_string (J.to_list (J.member "profiles" input))
              in
              Events.encode_hello
                Events.
                  {
                    versions;
                    profiles;
                    hello_service = opt_str input "service";
                    hello_auth = opt_str input "auth";
                  }
          | "hello_ack" ->
              Events.encode_hello_ack
                Events.
                  {
                    v = req_i64 input "v";
                    ack_profile = req_str input "profile";
                    session = opt_str input "session";
                  }
          | "ping" ->
              Events.encode_heartbeat
                Events.
                  { nonce = req_i64 input "nonce"; at = opt_i64 input "at" }
          | "close" ->
              Events.encode_close
                Events.
                  {
                    status =
                      Conventions.status_from_code (req_i64 input "status");
                    reason = opt_str input "reason";
                  }
          | other -> failwith ("unknown control " ^ other)
        in
        expect_hex name ~got:(tohex bytes) ~want:hex
      else
        let profile =
          match Events.parse_profile (req_str input "profile") with
          | Some p -> p
          | None -> failwith "unknown profile"
        in
        let payload = unhex (req_str input "payload_hex") in
        let ev =
          match profile with
          | Events.Verbose -> (
              let e =
                Events.new_verbose_event (opt_str input "service")
                  (req_str input "event") payload
              in
              match opt_i64 input "id" with
              | Some id -> Events.event_with_id e id
              | None -> e)
          | Events.Compact -> (
              let e =
                Events.new_compact_event
                  (req_i64 input "service_ord")
                  (req_i64 input "op_ord") payload
              in
              match opt_i64 input "id" with
              | Some id -> Events.event_with_id e id
              | None -> e)
        in
        match Events.encode_event ev profile with
        | Ok bytes -> (
            expect_hex name ~got:(tohex bytes) ~want:hex;
            match Events.decode_event bytes profile with
            | Ok dec -> expect (name ^ " decode") (dec = ev)
            | Error e -> die (name ^ " decode") e)
        | Error e -> die (name ^ " encode") e)
    (load "events.json")

let datagram_vectors () =
  List.iter
    (fun vec ->
      let name = J.to_string (J.member "name" vec) in
      let hex = J.to_string (J.member "hex" vec) in
      let input = J.member "input" vec in
      match req_str input "profile" with
      | "cbor-array" -> (
          let d =
            Datagrams.new_datagram (req_i64 input "op_ord")
              (req_i64 input "seq")
              (unhex (req_str input "payload_hex"))
          in
          let bytes = Datagrams.encode_datagram d in
          expect_hex name ~got:(tohex bytes) ~want:hex;
          match Datagrams.decode_datagram bytes with
          | Ok dec -> expect (name ^ " decode") (dec = d)
          | Error e -> die (name ^ " decode") e)
      | "compact-header" -> (
          let d =
            Datagrams.new_compact_datagram (req_int input "op_ord")
              (req_int input "seq")
              (unhex (req_str input "body_hex"))
          in
          let d =
            match opt_int input "epoch" with
            | Some e -> Datagrams.compact_with_epoch d e
            | None -> d
          in
          let bytes = Datagrams.encode_compact_datagram d in
          expect_hex name ~got:(tohex bytes) ~want:hex;
          match Datagrams.decode_compact_datagram bytes with
          | Ok dec -> expect (name ^ " decode") (dec = d)
          | Error e -> die (name ^ " decode") e)
      | other -> failwith ("unknown datagram profile " ^ other))
    (load "datagrams.json")

let run () =
  rpc_vectors ();
  events_vectors ();
  datagram_vectors ();
  print_endline "conformance: all vectors passed"
